pulsar-broker/src/test/java/org/apache/pulsar/client/api/DispatcherLockTest.java

```diff
@@ -0,0 +1,133 @@
+     * 1. Register "consumer-1"
+     *   - "consumer-1" will be maintained by the Subscription.
+     *   - "consumer-1" will be maintained by the Dispatcher.
+     * 2. The connection of "consumer-1" has something wrong. We call this connection "connection-1"
+     * 3. Try to register "consumer-2"
+     *   - "consumer-2" will be maintained by the Subscription. At this time, there are two consumers under this
+     *     subscription.
+     *   - This will trigger a connection check task for connection-1, we call this task "CheckConnectionLiveness".
+     *     This task will be executed in another thread, which means it will release the lock `Synchronized(dispatcher)`
+     *   - "consumer-2" was not maintained by the Dispatcher yet.
+     * 4. "CheckConnectionLiveness" will kick out "consumer-1" after 5 seconds, then "consumer-2" will be maintained
+     *    by the Dispatcher.
+     * (Highlight) Race condition: if the connection of "consumer-2" went to a wrong state before step 4,
+     *   "consumer-2" maintained by the Subscription and not maintained by the Dispatcher. Would the scenario below
+     *   will happen?
+     *   1. "connection-2" closed.
+     *   2. Remove "consumer-2" from the Subscription.
+     *   3. Try to remove "consumer-2" from the Dispatcher, but there are no consumers under this Dispatcher. To remove
+     *      nothing.
+     *   4. "CheckConnectionLiveness" is finished; put "consumer-2" into the Dispatcher.
+     *   5. At this moment, the consumer's state of Subscription and Dispatcher are not consistent. There is an orphan
+     *      consumer under the Dispatcher.
+     */
+    @Test
+    public void testNoOrphanConsumerIfLostDispatcherLock() throws Exception {
+        final String tpName = BrokerTestUtil.newUniqueName("persistent://public/default/tp");
+        final String subscription = "s1";
+        admin.topics().createNonPartitionedTopic(tpName);
+        admin.topics().createSubscription(tpName, subscription, MessageId.earliest);
+        List<Range> ranges = Collections.singletonList(new Range(0, 65535));
+        KeySharedPolicySticky sharedPolicySticky = new KeySharedPolicySticky.KeySharedPolicySticky().ranges(ranges);
+        final String consumerName1 = "c1";
+        final String consumerName2 = "c2";
+
+        // Create a client that injected logic: do not answer for the command Ping
+        ClientBuilderImpl clientBuilder = (ClientBuilderImpl) PulsarClient.builder().serviceUrl(lookupUrl.toString());
+        PulsarClient skipHealthCheckClient = InjectedClientCnxClientBuilder.create(clientBuilder,
+            new ClientCnxCustomizer(Type.PING){
+                @Override
+                public void handleCommand(Object command) {
+                    if (command instanceof CommandPing) {
+                        // do not response anything.
+                    }
+                }
+            });
```

Comment:
Is this ClientCnxCustomizer any easier than providing an actual anonymous ClientCnx class?

Passing a `BiFunction<ClientConfigurationData, EventLoopGroup, ClientCnx>` as a parameter in the `InjectedClientCnxClientBuilder` should be sufficient to support anything that is needed.

Reply:
I think your suggestion is better.

I created a new implementation [here](https://github.com/apache/pulsar/pull/21684/files#diff-ca20ea65d19296e57446d7f095cc00bcabb2dc3cb002d9c69d2b6d9f3c9f78e6), after the PR merged, I will back to finish this one.

Thanks

Reply:
Done